packages/shared_preferences/shared_preferences_android/android/src/test/java/io/flutter/plugins/sharedpreferences/SharedPreferencesTest.java

```diff
@@ -4,12 +4,131 @@
+import java.util.Map;
 import org.junit.Test;
 
+// Class Mocking SharedPreferencesAPI
+class MockAPI implements Messages.SharedPreferencesApi {
+
+  Map<String, Object> items = new HashMap<String, Object>();
+
+  @Override
+  public Boolean remove(@NonNull String key) {
+    items.remove(key);
+    return true;
+  }
+
+  @Override
+  public Boolean setBool(@NonNull String key, @NonNull Boolean value) {
+    items.put(key, value);
+    return true;
+  }
+
+  @Override
+  public Boolean setString(@NonNull String key, @NonNull String value) {
+    items.put(key, value);
+    return true;
+  }
+
+  @Override
+  public Boolean setInt(@NonNull String key, @NonNull Object value) {
+    items.put(key, value);
+    return true;
+  }
+
+  @Override
+  public Boolean setDouble(@NonNull String key, @NonNull Double value) {
+    items.put(key, value);
+    return true;
+  }
+
+  @Override
+  public Boolean setStringList(@NonNull String key, @NonNull List<String> value) {
+    items.put(key, value);
+    return true;
+  }
+
+  @Override
+  public Boolean clear() {
+    items.clear();
+    return true;
+  }
+
+  @Override
+  public Map<String, Object> getAll() {
+    return items;
+  }
+}
+
 public class SharedPreferencesTest {
-  // This is only a placeholder test and doesn't actually initialize the plugin.
+
+  MockAPI api = new MockAPI();
+
+  Map<String, Object> data =
+      new HashMap<String, Object>() {
+        {
+          put("Language", "Java");
+          put("Counter", 0);
+          put("Pie", 3.14);
+          put("Names", Arrays.asList("Flutter", "Dart"));
+          put("NewToFlutter", false);
+        }
+      };
+
   @Test
   public void initPluginDoesNotThrow() {
     final SharedPreferencesPlugin plugin = new SharedPreferencesPlugin();
   }
+
+  @Test
+  public void testAddValues() {
+    assertEquals(api.getAll().size(), 0);
+
+    addData(api);
+
+    assertEquals(api.getAll().size(), 5);
+    assertEquals(api.getAll().get("Language"), "Java");
+    assertEquals(api.getAll().get("Counter"), 0);
+    assertEquals(api.getAll().get("Pie"), 3.14);
+    assertEquals(api.getAll().get("Names"), Arrays.asList("Flutter", "Dart"));
+    assertEquals(api.getAll().get("NewToFlutter"), false);
+  }
+
+  @Test
+  public void testGetAllData() {
+    assertEquals(api.getAll(), new HashMap<String, Object>());
+    addData(api);
+
+    assertEquals(api.getAll(), data);
+  }
+
+  @Test
+  public void testClear() {
+    addData(api);
+    assertEquals(api.getAll().size(), 5);
+    api.clear();
+    assertEquals(api.getAll().size(), 0);
+  }
+
+  @Test
+  public void testRemove() {
+    api.setBool("isJava", true);
+    api.remove("isJava");
+    assertFalse(api.getAll().containsKey("isJava"));
+  }
+
+  public void addData(MockAPI api) {
+    api.setString("Language", "Java");
+    api.setInt("Counter", 0);
+    api.setDouble("Pie", 3.14);
+    api.setStringList("Names", Arrays.asList("Flutter", "Dart"));
+    api.setBool("NewToFlutter", false);
+  }
```

Comment:
I wrote these tests so that the user can easily understand what all functions are available and what all values can be passed. I am using Fake Tests because we can't access the `SharedPreferencesPlugin` in Unit Tests. If I am wrong about it, please let me know.

Reply:
> I wrote these tests so that the user can easily understand what all functions are available and what all values can be passed.

The purpose of unit tests is to test code; documentation belongs in comments.

> I am using Fake Tests because we can't access the `SharedPreferencesPlugin` in Unit Tests.

Why wouldn't unit tests be able to access code under test?

Reply:
> Because under `SharedPreferencesPlugin.java` we are using `android.content.SharedPreferences` and other libraries which can only execute when we test on an Emulator/device.

Internal implementation details that aren't available in unit tests can be mocked out, per my comment above.

> > I took inspiration for tests from [here](https://github.com/flutter/plugins/blob/main/packages/path_provider/path_provider_android/android/src/test/java/io/flutter/plugins/pathprovider/StorageDirectoryMapperTest.java).
> 
> Let me know if I'm wrong here.

That is testing a small utility, not the plugin itself, so isn't going to be very useful as an example. But even then, that's testing actual code, not testing test mocks, so is serving a useful test purpose.

Examples of substantive unit tests include [this](https://github.com/flutter/plugins/blob/main/packages/local_auth/local_auth_android/android/src/test/java/io/flutter/plugins/localauth/LocalAuthTest.java) and [this](https://github.com/flutter/plugins/blob/main/packages/url_launcher/url_launcher_android/android/src/test/java/io/flutter/plugins/urllauncher/MethodCallHandlerImplTest.java).

Reply:
Hey there, I went through the examples and I was able to create some tests using `MockedSharedPreferencesPlugin`. 

- I tried, mocking `android.content.SharedPreferences`: I tried this method but it results in a chain of mapping methods similar to the approach I have followed.
- I don't think we will be able to test the production code like the code under `MethodCallHandlerImpl` because it requires several mocked objects with chained methods `Mockito.when()` to map methods.

[Here](https://pastebin.com/RUa5r4yg), is the updated test. I am really unsure about how to write tests for this because there are not many references of tests with packages using `Pigeon`.